### PR TITLE
kill: adds support for numeric signals

### PR DESCRIPTION
**Pull request by devimc**
With this commit kill command is able to handle numeric
signals as argument

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

File: kill.go [hunk 1]
```diff
@@ -129,7 +130,27 @@ func kill(containerID, signal string, all bool) error {
...
+		s, err := strconv.Atoi(signal)
+		if err != nil {
+			return fmt.Errorf("Failed to convert signal %s to int", signal)
+		}
+		signum = syscall.Signal(s)
+		// check whether signal is valid or not
+		for _, v := range signals {
+			if v == signum {
+				// signal is a valid signal
+				if err := vc.KillContainer(containerID, podStatus.ContainersStatus[0].ID, signum); err != nil {
```
Comment (jodh-intel) [inline kill.go]: I'd prefer only a single call to `vc.KillContainer()` if possible.

Also, note #109 if you want to pick that up?

File: kill.go [hunk 2]
```diff
@@ -129,7 +130,27 @@ func kill(containerID, signal string, all bool) error {
 		return fmt.Errorf("Process not running inside container %s", containerID)
 	}
 
-	if err := vc.KillContainer(containerID, podStatus.ContainersStatus[0].ID, signals[signal]); err != nil {
+	signum, signalOk := signals[signal]
```
Comment (dlespiau) [inline kill.go]: You have the same code repeated twice now with some somewhat deep nested indentation.

```
				if err := vc.KillContainer(containerID, podStatus.ContainersStatus[0].ID, signum); err != nil {
					return err
				}
				return nil
```

1. resolve the argument to get a signal number
2. do the KillContainer call

For 1. we don't actually have to cycle through the map, we can just test if the signal is between 1 and 31

File: kill.go (hunk 1 above)
Comment (devimc) [inline kill.go]: yep, I'll take this issue

**Comment by devimc** [comment]
@jodh-intel @sboeuf  @dlespiau  Fixed
Thanks

File: kill_test.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"testing"
+)
+
+func TestProcessSignal(t *testing.T) {
```
Comment (dlespiau) [inline kill_test.go]: The usual thing here to limit the code to type and maintain is to use an anonymous struct: https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go

Comment (dlespiau) [inline kill_test.go]: something like

```
tests := []struct {
  input string
  valid bool
  golden
} {
 /* list of test cases */
}

for _, test := range tests {

 .....

}
```

An example at: https://github.com/dlespiau/x86db/blob/master/db_test.go#L10

**Comment by dlespiau** [comment]
with the list of test cases input in a struct this is:

lgtm

File: kill.go [hunk 4]
```diff
@@ -129,9 +130,42 @@ func kill(containerID, signal string, all bool) error {
...
+	signum, signalOk = signals["SIG"+signal]
+	if signalOk {
+		return signum, nil
+	}
+
+	// Support for numeric signals
+	s, err := strconv.Atoi(signal)
+	if err != nil {
+		return 0, fmt.Errorf("Failed to convert signal %s to int", signal)
+	}
```
Comment (sboeuf) [inline kill.go]: Add a new line please.

File: kill.go [hunk 5]
```diff
@@ -129,9 +130,42 @@ func kill(containerID, signal string, all bool) error {
...
+	signum = syscall.Signal(s)
+	// check whether signal is valid or not
```
Comment (sboeuf) [inline kill.go]: s/check/Check/

File: kill.go [hunk 6]
```diff
@@ -129,9 +130,42 @@ func kill(containerID, signal string, all bool) error {
...
+	for _, v := range signals {
+		if v == signum {
+			// signal is a valid signal
+			return signum, nil
+		}
+	}
```
Comment (sboeuf) [inline kill.go]: Add a new line.

File: kill.go [hunk 7]
```diff
@@ -129,9 +130,42 @@ func kill(containerID, signal string, all bool) error {
...
+
+	// Support for numeric signals
+	s, err := strconv.Atoi(signal)
+	if err != nil {
+		return 0, fmt.Errorf("Failed to convert signal %s to int", signal)
+	}
+	signum = syscall.Signal(s)
+	// check whether signal is valid or not
+	for _, v := range signals {
```
Comment (sboeuf) [inline kill.go]: Maybe use "sig" instead of "v" because this will be easier to understand.

File: kill_test.go (hunk 3 above)
Comment (sboeuf) [inline kill_test.go]: Or another way is to actually create each test such as `TestProcessSignalInvalidSignal()`, `TestProcessSignalInvalidShortSignal()`, ... into a generic function like `testProcessSignalFailure()` or `testProcessSignalSuccess()`.
That way each unit test has a name and it is very quick to get which test failed.

**Comment by sboeuf** [comment]
Oh BTW @devimc could you please rebase recent changes from github.com/clearcontainers/runtime:origin/master so that Semaphore CI build can pass.

File: kill_test.go (hunk 3 above)
Comment (dlespiau) [inline kill_test.go]: @sboeuf that adds quite a lot of typing and duplication actually. If the input needs some description, it can be added to the entries of the table, especially for very simple functions like that one that just take a string as input and return an int.

File: kill_test.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+		{"SIGDCKBY", 0}, //invalid signal
+		{"DCKBY", 0},    //invalid signal
+		{"99999", 0},    //invalid signal
+		{"SIGTERM", syscall.SIGTERM},
+		{"TERM", syscall.SIGTERM},
+		{"15", syscall.SIGTERM},
+	}
+
+	for _, test := range tests {
+		signum, _ := processSignal(test.signal)
```
Comment (dlespiau) [inline kill_test.go]: you're missing testing  we receive a non nil error on invalid input.

**Comment by devimc** [comment]
@dlespiau  @sboeuf  fixed

**Comment by dlespiau** [comment]
\o/
